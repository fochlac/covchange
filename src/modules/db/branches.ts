import { outputFile as outputFileRaw, readJSON as readJSONRaw } from 'fs-extra'

import { forge } from '../../utils/forge'
import initDb from '../../utils/fileDb'

const writeDiffs = (id, data) =>
	outputFileRaw(global.storage + `branches/branch_${id}.json`, typeof data !== 'string' ? JSON.stringify(data) : data, 'utf8')
const readDiffs = id => readJSONRaw(global.storage + `branches/branch_${id}.json`, 'utf8')

class BranchDb implements Core.BranchDb {
	db: Core.FileDb
	ready: Promise<void>

	constructor() {
		this.ready = initDb('branches.json').then((db: Core.FileDb) => {
			this.db = db
		})
		this.get = this.get.bind(this)
		this.create = this.create.bind(this)
		this.update = this.update.bind(this)
	}

	async exists(repository: Core.Repository, name: string): Promise<boolean> {
		await this.ready
		const id = `${repository}_${name}`
		return !!this.db.get(id)
	}

	async get(repository: Core.Repository, name: string): Promise<Core.Branch> {
		await this.ready
		const id = forge.branchId({ repository, name })
		const branch = this.db.get(id)
		if (branch) {
			const reports = await readDiffs(id)
			return { ...branch, reports }
		}
		return null
	}

	async update(branchRest: Core.BranchRest): Promise<Core.Branch> {
		await this.ready
		if (!(await this.exists(branchRest.repository, branchRest.name))) return Promise.reject('Branch does not exist')
		const id = forge.branchId(branchRest)
		const reports = (await readDiffs(id)).filter(({ timestamp }) => timestamp !== branchRest.report.timestamp)
		const branch = this.db.get(id)
		branch.lastModified = Date.now()
		await Promise.all([writeDiffs(id, [branchRest.report, ...reports]), this.db.set(id, branch)])
		branch.reports = [branchRest.report, ...reports]

		return branch
	}

	async create(branchRest: Core.BranchRest): Promise<Core.Branch> {
		await this.ready
		if (await this.exists(branchRest.repository, branchRest.name)) return Promise.reject('Branch already exists')
		const id = forge.branchId(branchRest)
		const branch = { ...branchRest, id, lastModified: Date.now() } as Core.Branch
		await Promise.all([writeDiffs(id, [branchRest.report]), this.db.set(id, branch)])
		branch.reports = [branchRest.report]
		return branch
	}
}

export const branchDb = new BranchDb()
