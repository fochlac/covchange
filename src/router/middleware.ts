import { createReportFromXML } from '../utils/parsing/parse-report'

export const parseReport = (req, res, next) => {
	req.body.report = createReportFromXML(req.body.report)
	next()
}
