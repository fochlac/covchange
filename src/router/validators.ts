import { validate as isValidXML } from 'fast-xml-parser'

export const branch = /^[^\s]{5,250}$/
export const repository = {
	repo: /^[^\s]{5,250}$/,
	type: /^(users|projects)$/,
	project: /^[a-zA-Z-_]{5,250}$/,
}
export const pullRequestId = /^[\n]{5,10}$/
export const url = /^(?:http(s)?:\/\/)?[\w.-]+(?:\.[\w\.-]+)+[\w\-\._~:/?#[\]@!\$&'\(\)\*\+,;=.]+$/
// @ts-ignore
url.allowUndefined = true

export const report = input => {
	return !!input && isValidXML(input, {}) === true
}
