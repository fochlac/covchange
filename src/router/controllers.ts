import { branchDb } from '../modules/db/branches'
import error from '../utils/error'
import { getBaseBranchFromPullrequest } from '../modules/bitbucket/get-base-branch'
import { pullRequestDb } from '../modules/db/pull-requests'
import { writeReportToBitbucket } from '../modules/bitbucket/comment'

const { routerError, internalError } = error('controllers:')

export const addBranchReport = async (req, res) => {
	const { name, repository, report } = req.body
	const action = (await branchDb.exists(repository, name)) ? branchDb.update : branchDb.create

	action({ name, repository, report })
		.then(() => {
			res.status(200).send({ success: true })
		})
		.catch(routerError(2, res, 'Error adding branch report.'))
}

export const addPullRequestReport = async (req, res) => {
	try {
		const pullRequestRest: Core.PullRequestRest = { ...req.body, task: !!req.body.task }
		const action = (await pullRequestDb.exists(pullRequestRest.repository, pullRequestRest.name))
			? pullRequestDb.update
			: pullRequestDb.create
		const base = await getBaseBranchFromPullrequest(pullRequestRest)

		const baseBranch = await branchDb.get(base.repository, base.name)
		if (!baseBranch) throw 'Base branch is not in the database.'
		const pr = await action(pullRequestRest, base)

		await writeReportToBitbucket(baseBranch, pr).catch(internalError(2, 'Error handling pull request:'))
		res.status(200).send({ success: true })
	} catch (err) {
		routerError(2, res, 'Error adding pullrequest report.')(err)
	}
}
