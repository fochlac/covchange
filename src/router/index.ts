import { addBranchReport, addPullRequestReport } from './controllers'
import { branch, pullRequestId, report, repository } from './validators'

import { createRouter } from 'abstract-express-router'
import { join } from 'path'
import { parseReport } from './middleware'

export const router = createRouter(
	{
		api: {
			middleware: [],
			branch: {
				post: {
					body: { name: branch, report, repository },
					middleware: [parseReport],
					controller: addBranchReport,
				},
			},
			pullrequest: {
				post: {
					body: {
						name: pullRequestId,
						report,
						repository,
					},
					middleware: [parseReport],
					controller: addPullRequestReport,
				},
			},
		},
		svg: {
			static: join(__dirname, '../../svg/'),
		},
	},
	{
		logLevel: 3,
	},
)
