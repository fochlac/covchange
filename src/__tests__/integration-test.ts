import './mock-config'

import { mockBranchRest } from './data/branches'
import { mockPullRequestRest } from './data/pullrequests'
import nock from 'nock'
import { reportRaw } from './data/reports'
import { router } from '../router'
import supertest from 'supertest'

jest.mock('../utils/file-db')
jest.mock('../utils/logger')
jest.mock('fs-extra', () => {
	let data = {}

	return {
		outputFile: jest.fn((file, content) => {
			data[file] = content
			return Promise.resolve()
		}),
		readJSON: jest.fn(file => Promise.resolve(JSON.parse(data[file]))),
	}
})

describe('upload branch, upload pr, expect repo', () => {
	it('should send svgs', async () => {
		const response = await supertest(router).get('/svg/arrow-up.svg')

		expect(response.statusCode).toBe(200)
		expect(response).toMatchObject({
			header: {
				'accept-ranges': 'bytes',
				'cache-control': 'public, max-age=0',
				connection: 'close',
				'content-length': '547',
				'content-type': 'image/svg+xml',
				vary: 'Accept-Encoding',
				'x-powered-by': 'Express',
			},
		})
	})

	it('should create branch, create pullrequest and send comment', async () => {
		const bitbucket = nock('http://test.bitbucket.server:8080')
			.get('/rest/api/1.0/users/slug/repos/name/pull-requests/12345/')
			.reply('200', {
				toRef: {
					displayId: mockBranchRest.name,
					repository: {
						slug: mockBranchRest.repository.repo,
						project: {
							key: mockBranchRest.repository.project,
							type: 'USERS',
						},
					},
				},
			})
			.post('/rest/api/1.0/users/slug/repos/name/pull-requests/12345/comments/')
			.reply(200, { version: 0, id: 1 })

		await supertest(router)
			.post('/api/branch')
			.send({ ...mockBranchRest, report: reportRaw })
			.expect(200)

		await supertest(router)
			.post('/api/pullrequest')
			.send({ ...mockPullRequestRest, report: reportRaw })
			.expect(200)

		expect(bitbucket.isDone()).toBeTruthy()
	})

	it('should update pullrequest and update comment', async () => {
		const bitbucket = nock('http://test.bitbucket.server:8080')
			.get('/rest/api/1.0/users/slug/repos/name/pull-requests/12345/')
			.reply('200', {
				toRef: {
					displayId: mockBranchRest.name,
					repository: {
						slug: mockBranchRest.repository.repo,
						project: {
							key: mockBranchRest.repository.project,
							type: 'USERS',
						},
					},
				},
			})
			.get('/rest/api/1.0/users/slug/repos/name/pull-requests/12345/comments/1')
			.reply(200, { version: 2, id: 1 })
		.put('/rest/api/1.0/users/slug/repos/name/pull-requests/12345/comments/1', { version: 2, text: /.*/})
			.reply(200, { version: 3, id: 1 })

		await supertest(router)
			.post('/api/pullrequest')
			.send({ ...mockPullRequestRest, report: reportRaw })
			.expect(200)

		expect(bitbucket.isDone()).toBeTruthy()
	})
})
