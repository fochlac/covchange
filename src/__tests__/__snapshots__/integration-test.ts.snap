// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`upload branch, upload pr, expect repo should create branch, create pullrequest and send comment 1`] = `
Array [
  Array [
    8,
    " - router - ",
    "POST call to /api/branch",
  ],
  Array [
    6,
    " - router - ",
    "Validation succeded for call to /api/branch",
  ],
  Array [
    8,
    " - router - ",
    "POST call to /api/pullrequest",
  ],
  Array [
    6,
    " - router - ",
    "Validation succeded for call to /api/pullrequest",
  ],
]
`;

exports[`upload branch, upload pr, expect repo should send svgs 1`] = `
Array [
  Array [
    8,
    " - router - ",
    "GET call to /svg/arrow-up.svg",
  ],
]
`;

exports[`upload branch, upload pr, expect repo should update branch, update comment and create task 1`] = `
Array [
  Array [
    8,
    " - router - ",
    "POST call to /api/branch",
  ],
  Array [
    6,
    " - router - ",
    "Validation succeded for call to /api/branch",
  ],
  Array [
    8,
    " - router - ",
    "POST call to /api/pullrequest",
  ],
  Array [
    6,
    " - router - ",
    "Validation succeded for call to /api/pullrequest",
  ],
]
`;

exports[`upload branch, upload pr, expect repo should update pullrequest and update comment 1`] = `
Array [
  Array [
    8,
    " - router - ",
    "POST call to /api/pullrequest",
  ],
  Array [
    6,
    " - router - ",
    "Validation succeded for call to /api/pullrequest",
  ],
]
`;
